src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+                        return ((Http3FrameType)frameType, payloadLength);
+                    case Http3FrameType.Settings:
+                    case Http3FrameType.GoAway:
+                    case Http3FrameType.MaxPushId:
+                        // These frames should only be received on a control stream, not a response stream.
+                        throw new Http3ConnectionException(Http3ErrorCode.UnexpectedFrame);
+                    case Http3FrameType.DuplicatePush:
+                    case Http3FrameType.PushPromise:
+                    case Http3FrameType.CancelPush:
+                        // Because we haven't sent any MAX_PUSH_ID frames, any of these push-related
+                        // frames that the server sends will have an out-of-range push ID.
+                        // TODO: close the conenction with H3_ID_ERROR.
+                        throw new Http3ConnectionException(Http3ErrorCode.IdError);
+                    default:
+                        // Unknown frame types should be skipped.
+                        await SkipUnknownPayloadAsync((Http3FrameType)frameType, payloadLength, cancellationToken).ConfigureAwait(false);
+                        break;
+                }
+            }
+        }
+
+        private async ValueTask ReadHeadersAsync(long headersLength, CancellationToken cancellationToken)
+        {
+            if (headersLength > _headerBudgetRemaining)
+            {
+                throw new HttpRequestException(SR.Format(SR.net_http_response_headers_exceeded_length, _connection.Pool.Settings._maxResponseHeadersLength * 1024L));
+            }
+
+            _headerBudgetRemaining -= headersLength;
+
+            while (headersLength != 0)
+            {
+                if (_buffer.ActiveLength == 0)
+                {
+                    _buffer.EnsureAvailableSpace(1);
+
+                    int bytesRead = await _stream.ReadAsync(_buffer.AvailableMemory, cancellationToken).ConfigureAwait(false);
+                    if (bytesRead != 0)
+                    {
+                        _buffer.Commit(bytesRead);
+                    }
+                    else
+                    {
+                        throw new HttpRequestException("Partial frame received; unexpected end of stream.");
+                    }
+                }
+
+                int processLength = (int)Math.Min(headersLength, _buffer.ActiveLength);
+
+                _headerDecoder.Decode(_buffer.ActiveSpan.Slice(processLength), this);
+                _buffer.Discard(processLength);
+                headersLength -= processLength;
+            }
+        }
+
+        private static ReadOnlySpan<byte> StatusHeaderNameBytes => new byte[] { (byte)'s', (byte)'t', (byte)'a', (byte)'t', (byte)'u', (byte)'s' };
+
+        void IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+        {
+            Debug.Assert(name.Length > 0);
+
+            if (name[0] == ':')
```

Comment:
Is the first call to OnHeader always guaranteed to start with :status?

Reply:
I think you may hit a null ref on response if not.

Reply:
TBD; I thought this was weird too, but it appears to be the way our HTTP/2 implementation works, so I copied that behavior for now. Will ask around.

Reply:
Confirmed this is correct behavior.

H3 says:
> The restrictions on the use of pseudo-header fields in Section 8.1.2 of HTTP2 also apply to HTTP/3

That section of H2 says:
> All pseudo-header fields MUST appear in the header block before regular header fields.  Any request or response that contains a pseudo-header field that appears in a header block after a regular header field MUST be treated as malformed